source/ConditionsStore.h

```diff
@@ -0,0 +1,152 @@
+		const std::pair<const std::string, int64_t>*, // iterator: pointer
+		std::pair<const std::string, int64_t>>        // iterator: reference
+	{
+		using CondMapItType = std::map<std::string, ConditionEntry>::const_iterator;
+	
+	public:
+		PrimariesIterator(CondMapItType it, CondMapItType endIt);
+		
+		// Default input_iterator operations.
+		std::pair<std::string, int64_t> operator*() const;
+		const std::pair<std::string, int64_t>* operator->();
+		PrimariesIterator& operator++();
+		PrimariesIterator operator++(int);
+		bool operator== (const PrimariesIterator& rhs) const;
+		bool operator!= (const PrimariesIterator& rhs) const;
+	
+	
+	private:
+		// Helper function to ensure that the primary-conditions iterator points
+		// to a primary (value) condition or to the end-iterator value.
+		void MoveToValueCondition();
+	
+	
+	private:
+		// The operator->() requires a return value that is a pointer, but in this
+		// case there is no original pair-object to point to, so we generate a
+		// virtual object on the fly while iterating.
+		std::pair<std::string, int64_t> itVal;
+		CondMapItType condMapIt;
+		CondMapItType condMapEnd;
+	};
+
+
+
+public:
+	// Constructors to initialize this class.
+	ConditionsStore() = default;
+	ConditionsStore(std::initializer_list<std::pair<std::string, int64_t>> initialConditions);
+	ConditionsStore(const std::map<std::string, int64_t> &initialConditions);
+
+	// Retrieve a "condition" flag from this store (directly or from the
+	// connected provider).
+	int64_t GetCondition(const std::string &name) const;
+	bool HasCondition(const std::string &name) const;
+	// Add a value to a condition, set a value for a condition or erase a
+	// condition completely. Returns true on success, false on failure.
+	bool AddCondition(const std::string &name, int64_t value);
+	bool SetCondition(const std::string &name, int64_t value);
+	bool EraseCondition(const std::string &name);
+	
+	// Direct (read-only) access to the stored primary conditions.
+	PrimariesIterator PrimariesBegin() const;
+	PrimariesIterator PrimariesEnd() const;
+	PrimariesIterator PrimariesLowerBound(const std::string &key) const;
+	
+	// Set providers for derived conditions based on prefix and name.
+	void SetProviderPrefixed(const std::string &prefix, DerivedProvider conditionsProvider);
+	void SetProviderNamed(const std::string &name, DerivedProvider conditionProvider);
+	
+	// Helper to completely remove all data and linked condition-providers from the store.
+	void Clear();
+
+
+private:
+	// Retrieve a condition entry based on a condition name, doCreate indicate if the entry
+	// should be created if it wasn't there yet.
+	ConditionEntry *GetEntry(const std::string &name);
+	const ConditionEntry *GetEntryConst(const std::string &name) const;
+
+
+
+private:
+	// Storage for both the primary conditions as well as the providers.
+	std::map<std::string, ConditionEntry> storage;
```

Comment:
Quick question: Have you considered having 3 maps instead of one? It might simplify lots of code (like the iterator not being needed anymore). We'll need to do 3 map lookups in the worst case to get a condition though but map lookups aren't expensive. I haven't fully thought it out so just a thought.

Reply:
Yes, earlier implementations had 2 or 3 separate maps, but tehhowch was concerned about the performance when we would need to do all those separate lookups, so I switched to this (more complex) implementation that gets the relevant condition(provider) in a single lookup.

Reply:
I guess he has a point. Although IMO: The majority of condition access are for value conditions right? Sooo you'd only have multiple lookups for anything other than a value condition....... :)

Reply:
Derived conditions take precedence over value conditions, so you would have 2 look-ups for value conditions:
-1- Check if a derived condition exists with the given name or prefix.
-2- Get the relevant value condition.

I guess we could do something smart to actively prune derived conditions out of the value conditions, but even then we would need 2 lookups for an unset value condition:
-1- Check if the value exists in the list of value conditions.
-2- If it is not set, then check if a derived condition exists.